# ***✅ Step 1: Extract and Inspect the Dataset***

In [ ]:
import zipfile
import os

# Define paths
zip_path = "D:\\Projects\\Guvi_Project5\\SolarGuard\\Faulty_solar_panel-20250430T004242Z-001.zip"
extract_path = "D:\\Projects\\Guvi_Project5\\SolarGuard\\data\\faulty_solar_panel_dataset"

# Extract the zip file
with zipfile.ZipFile(zip_path, 'r') as zip_ref:
    zip_ref.extractall(extract_path)

# Walk through the extracted directory and list the structure
dir_structure = {}
for root, dirs, files in os.walk(extract_path):
    relative_root = os.path.relpath(root, extract_path)
    dir_structure[relative_root] = {
        "dirs": dirs,
        "files": [f for f in files if f.lower().endswith(('.jpg', '.png', '.jpeg'))]
    }

dir_structure


{'.': {'dirs': ['Faulty_solar_panel'], 'files': []},
 'Faulty_solar_panel': {'dirs': ['Bird-drop',
   'Clean',
   'Dusty',
   'Electrical-damage',
   'Physical-Damage',
   'Snow-Covered'],
  'files': []},
 'Faulty_solar_panel\\Bird-drop': {'dirs': ['New'],
  'files': ['Bird (1).jpeg',
   'Bird (10).jpg',
   'Bird (100).jpg',
   'Bird (101).jpg',
   'Bird (102).jpg',
   'Bird (103).jpg',
   'Bird (104).jpg',
   'Bird (105).jpg',
   'Bird (106).jpg',
   'Bird (107).jpg',
   'Bird (108).jpg',
   'Bird (109).jpg',
   'Bird (11).jpg',
   'Bird (110).jpg',
   'Bird (111).jpg',
   'Bird (112).jpg',
   'Bird (113).jpg',
   'Bird (114).jpg',
   'Bird (115).jpg',
   'Bird (116).jpg',
   'Bird (117).jpg',
   'Bird (118).jpg',
   'Bird (119).jpg',
   'Bird (12).jpg',
   'Bird (120).jpg',
   'Bird (121).jpg',
   'Bird (122).jpg',
   'Bird (123).jpg',
   'Bird (124).jpg',
   'Bird (125).jpg',
   'Bird (126).jpg',
   'Bird (127).jpg',
   'Bird (128).jpg',
   'Bird (129).jpg',
   'Bird (13).jpg',
   '

## **📊 Processed Dataset Summary**

| Class               | Images |
|--------------------|--------|
| Clean              | 193    |
| Dusty              | 190    |
| Bird-drop          | 206    |
| Electrical-damage  | 99     |
| Physical-Damage    | 69     |
| Snow-Covered       | 123    |

In [ ]:
import shutil
from PIL import Image

# Output directory for processed images
processed_dir = "D:\\Projects\\Guvi_Project5\\SolarGuard\\data\\processed"
os.makedirs(processed_dir, exist_ok=True)

# Target image size
img_size = (224, 224)

def normalize_and_save_image(src_path, dst_path):
    try:
        img = Image.open(src_path).convert("RGB")
        img = img.resize(img_size)
        img.save(dst_path, format="JPEG")
    except Exception as e:
        print(f"Failed to process {src_path}: {e}")

# Process each class folder
src_base = os.path.join(extract_path, "Faulty_solar_panel")

for class_folder in os.listdir(src_base):
    class_path = os.path.join(src_base, class_folder)
    if not os.path.isdir(class_path):
        continue

    dst_class_path = os.path.join(processed_dir, class_folder.replace(" ", "_"))
    os.makedirs(dst_class_path, exist_ok=True)

    for root, _, files in os.walk(class_path):
        for file in files:
            if file.lower().endswith(('.jpg', '.jpeg', '.png')):
                src_file = os.path.join(root, file)
                dst_file = os.path.join(dst_class_path, os.path.splitext(file)[0] + ".jpg")
                normalize_and_save_image(src_file, dst_file)

# Verify result
processed_summary = {
    cls: len(files)
    for cls in os.listdir(processed_dir)
    if (files := os.listdir(os.path.join(processed_dir, cls)))
}

processed_summary


{'Bird-drop': 206,
 'Clean': 193,
 'Dusty': 190,
 'Electrical-damage': 99,
 'Physical-Damage': 69,
 'Snow-Covered': 123}

# ***✅ Step 2: Train/Test Split + Data Augmentation***

In [ ]:
import random
from shutil import copy2

# Define paths
final_base = "D:\\Projects\\Guvi_Project5\\SolarGuard\\data\\final"
train_dir = os.path.join(final_base, "train")
val_dir = os.path.join(final_base, "val")
os.makedirs(train_dir, exist_ok=True)
os.makedirs(val_dir, exist_ok=True)

split_ratio = 0.8  # 80% training, 20% validation

# Create train/val split for each class
for class_name in os.listdir(processed_dir):
    class_path = os.path.join(processed_dir, class_name)
    images = [f for f in os.listdir(class_path) if f.endswith(".jpg")]
    random.shuffle(images)

    split_idx = int(len(images) * split_ratio)
    train_imgs, val_imgs = images[:split_idx], images[split_idx:]

    # Create class folders
    train_class_dir = os.path.join(train_dir, class_name)
    val_class_dir = os.path.join(val_dir, class_name)
    os.makedirs(train_class_dir, exist_ok=True)
    os.makedirs(val_class_dir, exist_ok=True)

    # Copy images
    for img in train_imgs:
        copy2(os.path.join(class_path, img), os.path.join(train_class_dir, img))
    for img in val_imgs:
        copy2(os.path.join(class_path, img), os.path.join(val_class_dir, img))

# Summary of split
train_summary = {cls: len(os.listdir(os.path.join(train_dir, cls))) for cls in os.listdir(train_dir)}
val_summary = {cls: len(os.listdir(os.path.join(val_dir, cls))) for cls in os.listdir(val_dir)}

train_summary, val_summary


({'Bird-drop': 164,
  'Clean': 154,
  'Dusty': 152,
  'Electrical-damage': 79,
  'Physical-Damage': 55,
  'Snow-Covered': 98},
 {'Bird-drop': 42,
  'Clean': 39,
  'Dusty': 38,
  'Electrical-damage': 20,
  'Physical-Damage': 14,
  'Snow-Covered': 25})

# ***✅ Step 3:Train a CNN Model***

## ***We trained CNN models for panel condition classification***

1. ResNet50  
2. MobileNetV2  
3. EfficientNetB0  

### ***(02_train_all_models.py)✅ Training Completed for All 3 Models*** 


### ***📊 Model Training Results***

| Model           | Final Val Accuracy | Notes |
|----------------|--------------------|-------|
| **ResNet50**      | ~39%                | Struggled with convergence, underfitting likely due to insufficient features or frozen layers. |
| **MobileNetV2**   | **~80%**            | Best performance among all – good convergence and generalization. |
| **EfficientNetB0**| ~29%                | Underperformed significantly; may benefit from hyperparameter tuning or unfreezing. |

### ***🧠 Observations & Recommendations***  

✅ MobileNetV2 is currently your best classifier. It’s fast, accurate, and well-suited for edge deployment or real-time use.  

❗ ResNet50 and EfficientNetB0 are likely underfitting due to being frozen (trainable = False). You can try fine-tuning by unfreezing the top layers.  

📉 EfficientNetB0 may also require different learning rate or more training epochs.

# ***✅ Step 4:Evaluate_model.py – Evaluation***  

We evaluate the trained models (especially MobileNetV2) using:

1. Confusion Matrix

2. Classification Report (Precision, Recall, F1-score per class)

# ***📊 Classification Report Summary***

| Class              | Precision | Recall | F1-Score | Support |
|-------------------|-----------|--------|----------|---------|
| Bird-drop         | 0.70      | 0.79   | 0.74     | 42      |
| Clean             | 0.75      | 0.92   | 0.83     | 39      |
| Dusty             | 0.77      | 0.63   | 0.70     | 38      |
| Electrical-damage | 0.87      | 0.65   | 0.74     | 20      |
| Physical-Damage   | 0.75      | 0.64   | 0.69     | 14      |
| Snow-Covered      | 0.88      | 0.88   | 0.88     | 25      |
| **Overall Accuracy** |         |        | **0.77**  | 178     |

### **🔍 Insights**  

1. Clean and Snow-Covered classes are predicted most reliably.  

2. Electrical-damage has excellent precision but could benefit from recall improvement.  

3. Dusty and Physical-Damage are the weakest—likely due to:  

    *. Visual similarity to other classes  

    *. Fewer distinct features or sample imbalance  

# ***✅ Step 4: Object Detection*** (04_train_yolov8.py)

### ***Train models like YOLOv8 or Faster R-CNN to detect and localize damages/obstructions.***

### ***Creating YOLOv8 Annotations***
### **Prepare bounding box labels for:**  

1. Dusty

2. Bird-drop  

3. Electrical-damage  

4. Physical-damage  

5. Snow-Covered  

### **You do not need to annotate:**  

1. Clean  
---

## ✅ Here’s What to Do in Roboflow

### 📝 **1. Set Up Your Project**
- Go to [https://roboflow.com](https://roboflow.com)
- Create a free account (if you haven’t)
- Click **“Create New Project”**
  - Project Type: **Object Detection**
  - Project Name: `SolarGuard`
  - Classes: `Dust`, `Bird-drop`, `Electrical-damage`, `Physical-damage`, 'Snow-Covered'

---

### 📤 **2. Upload and Annotate Images**
- Upload images from your folders:
  - You can select from the `Dusty`, `Bird-drop`, `Electrical-damage`, `Physical-Damage`, 'Snow-Covered' categories.
- Start **drawing bounding boxes** for each object.
  - Make sure to label them accurately using the class dropdown.

---

### 📦 **3. Generate Dataset Version**
- Click **“Generate”** after annotating enough images.
- Choose a **train/val split** (e.g., 80/20)

---

### 📁 **4. Export in YOLOv5 Format**
- Click **Export Dataset**
  - Format: `YOLOv5 PyTorch`
  - Check: Include `data.yaml`
- Download the **ZIP file**

---

### 📥 **5. Upload ZIP Here**
Once exported, send me the ZIP file of your YOLO dataset (you can upload it here), and I’ll help you:

- ✅ Validate the structure
- ✅ Train the YOLOv8 model
- ✅ Optionally evaluate and visualize results

---

In [ ]:
import zipfile
import os

# Redefine new zip and extract path
zip_path = "D:\\Projects\\Guvi_Project5\\SolarGuard\\Object Detection.v3i.yolov5pytorch.zip"
extract_path = "D:\\Projects\\Guvi_Project5\\SolarGuard\\data\\final_yolo_dataset"

# Extract the new ZIP
with zipfile.ZipFile(zip_path, 'r') as zip_ref:
    zip_ref.extractall(extract_path)

# Check for data.yaml and class list
yaml_path = os.path.join(extract_path, "data.yaml")
yaml_exists = os.path.exists(yaml_path)

# Parse class names from YAML
classes = []
if yaml_exists:
    with open(yaml_path, 'r') as f:
        for line in f:
            if line.strip().startswith('names:'):
                class_list_line = line.strip()
                break
        else:
            class_list_line = None
    if class_list_line:
        try:
            classes = eval(class_list_line.split(":", 1)[1].strip())
        except Exception:
            pass

# Structure summary
structure = {}
for split in ["train", "valid", "test"]:
    images_dir = os.path.join(extract_path, split, "images")
    labels_dir = os.path.join(extract_path, split, "labels")
    images = os.listdir(images_dir) if os.path.exists(images_dir) else []
    labels = os.listdir(labels_dir) if os.path.exists(labels_dir) else []
    structure[split] = {
        "images": len(images),
        "labels": len(labels)
    }

{
    "yaml_found": yaml_exists,
    "classes": classes,
    "structure_summary": structure
}


{'yaml_found': True,
 'classes': ['Bird-drop',
  'Dusty',
  'Electrical-damage',
  'Physical-Damage',
  'Snow-Covered'],
 'structure_summary': {'train': {'images': 424, 'labels': 424},
  'valid': {'images': 113, 'labels': 113},
  'test': {'images': 61, 'labels': 61}}}

We've now uploaded the YOLO dataset:  
📦 `Object Detection.v3i.yolov5pytorch.zip`

✅ **Dataset Validated Successfully**

Our latest upload is perfect and ready for YOLOv8 training!

---

### 📋 Dataset Summary

- **Classes**:
  ```
  ['Bird-drop', 'Dusty', 'Electrical-damage', 'Physical-Damage', 'Snow-Covered']
  ```

- **Structure**:
  | Split  | Images | Labels |
  |--------|--------|--------|
  | Train  | 424    | 424    |
  | Valid  | 113    | 113    |
  | Test   | 61     | 61     |

- ✅ `data.yaml` is present and well-formed  
- ✅ Each image has a matching label  
- ✅ Class names are correctly mapped  

---

# ***✅ Step-5: Evaluation & Visualization***  

**evaluate_and_visualize.py and evaluate_metrics.py**- We evaluate our trained YOLOv8 model (best.pt) on the test set and visualize predictions with bounding boxes.  


📁 Make sure:  

Our trained model is here:  
runs/detect/solar_guard_detector2/weights/best.pt  

Our test images are here:  
final_yolo_dataset/test/images/  

![alt text](runs/detect/solar_guard_detector3/confusion_matrix_normalized.png)

![alt text](runs/detect/solar_guard_detector3/confusion_matrix.png)

![alt text](runs/detect/solar_guard_detector3/F1_curve.png)

![alt text](runs/detect/solar_guard_detector3/labels_correlogram.jpg)

![alt text](runs/detect/solar_guard_detector3/labels.jpg)

![alt text](runs/detect/solar_guard_detector3/P_curve.png)

![alt text](runs/detect/solar_guard_detector3/PR_curve.png)

![alt text](runs/detect/solar_guard_detector3/R_curve.png)

![alt text](runs/detect/solar_guard_detector3/results.png)

# ***✅ Step-6: Streamlit app.py**  

## **Features Included**  
**🖼️ Real-time Predictions:**
- Image classification (CNN – MobileNetV2)  

- Object detection (YOLOv8)  

- Maintenance recommendations based on detected condition  

**📊 EDA & Visualization:**  
- Class distribution chart from dataset  

- YOLOv8 training curves & confusion matrix  

**📈 Performance Metrics:**  
- Accuracy, Precision, Recall, F1, mAP (auto-generated by YOLOv8)  

**💡 Business Insights:**  
- Common defects  

- Preventive actions  

- Maintenance triggers